frame/contracts/src/lib.rs

```diff
@@ -1054,19 +1043,94 @@ impl Default for Schedule {
 	}
 }
 
+#[doc(hidden)]
+#[cfg_attr(test, derive(Debug))]
+pub struct DynamicWeightData<AccountId, NegativeImbalance> {
+	/// The account id who should receive the refund from the gas leftovers.
+	transactor: AccountId,
+	/// The negative imbalance obtained by withdrawing the value up to the requested gas limit.
+	imbalance: NegativeImbalance,
+}
+
 /// `SignedExtension` that checks if a transaction would exhausts the block gas limit.
 #[derive(Encode, Decode, Clone, Eq, PartialEq)]
 pub struct CheckBlockGasLimit<T: Trait + Send + Sync>(PhantomData<T>);
 
+impl<T: Trait + Send + Sync> CheckBlockGasLimit<T> {
+	/// Perform pre-dispatch checks for the given call and origin.
+	fn perform_pre_dispatch_checks(
+		who: &T::AccountId,
+		call: &<T as Trait>::Call,
+	) -> Result<Option<DynamicWeightData<T::AccountId, NegativeImbalanceOf<T>>>, TransactionValidityError> {
+		let call = match call.is_sub_type() {
+			Some(call) => call,
+			None => return Ok(None),
+		};
+
+		match *call {
+			Call::claim_surcharge(_, _) | Call::update_schedule(_) | Call::put_code(_) => Ok(None),
+			Call::call(_, _, gas_limit, _) | Call::instantiate(_, gas_limit, _, _) => {
+				// Compute how much block weight this transaction can take at its limit, i.e.
+				// if this transaction depleted all provided gas to zero.
+				let gas_weight_limit = Weight::try_from(gas_limit)
+					.map_err(|_| InvalidTransaction::ExhaustsResources)?;
+				let weight_available = <T as system::Trait>::MaximumBlockWeight::get()
+					.saturating_sub(<system::Module<T>>::all_extrinsics_weight()).into();
+				if gas_weight_limit > weight_available {
+					// We discard the transaction if the requested limit exceeds the available
+					// amount of weight in the current block.
+					//
+					// Note that this transaction is left out only from the current block and txq
+					// might attempt to include this transaction again.
+					return Err(InvalidTransaction::ExhaustsResources.into());
+				}
+
+				// Compute the fee corresponding for the given gas_weight_limit and attempt
+				// withdrawing from the origin of this transaction.
+				let fee = T::WeightToFee::convert(gas_weight_limit);
+
+				// Compute and store the effective price per unit of gas.
+				let gas_price = {
+					let divisor = gas_weight_limit.unique_saturated_into();
+					fee
+						.checked_div(&divisor)
+						.unwrap_or(1.into())
```

Comment:
this ratio can go very inaccurate if WeightToFee ratio is less than 1 or if it is different from a natural number. (i.e. 1.5)
Maybe we could make use of some Perbill here, maybe we should introduce in sp_arithmetic a perbill with a natural part as well `(1.5 = 1 + 500_000_000/1_000_000_000)`

Reply:
I think that shouldn't matter if the 1 weight = 1 gas, right? Apart from that, In upcoming https://github.com/paritytech/substrate/pull/4590 there are two instances of using gas_price:

1. `approx_gas_for_balance` to compute gas fee just before calling into the runtime. This should be changed with direct weight payments in dynamic weight world.
2. `approx_gas_for_balance` for getting gas amount for `TransferFeeToken`. #4590 [will use](https://github.com/paritytech/substrate/pull/4649/files#diff-7d69b0e04bf2c554f4fd753326b32a64R576) gas cost for the transfer. I believe we can change others as well.
3. Apart from `approx_gas_for_balance` we also have `ext_gas_price`, which I think we might want to drop.

So I am wondering if a TODO here would do?